src/main/java/net/neoforged/neoforge/common/crafting/outgredient/DefaultedItemTagOutgredient.java

```diff
@@ -0,0 +1,123 @@
+            Item.STREAM_CODEC.apply(ByteBufCodecs::optional), it -> it.fallback,
+            ByteBufCodecs.VAR_INT, it -> it.count,
+            DataComponentPatch.STREAM_CODEC, it -> it.components,
+            DefaultedItemTagOutgredient::new);
+
+    /**
+     * @param tagKey     The {@link TagKey} to use for looking up the outgredient.
+     * @param fallback   The fallback to use if the tag-based lookup did not yield a conclusive outgredient.
+     * @param count      The count to use. Corresponds to {@link ItemStack#getCount()}.
+     * @param components The data components to use. Corresponds to {@link ItemStack#getComponents()}.
+     */
+    public DefaultedItemTagOutgredient(TagKey<Item> tagKey, Holder<Item> fallback, int count, DataComponentPatch components) {
+        this(tagKey, Optional.of(fallback), count, components);
+    }
+
+    /**
+     * @param tagKey   The {@link TagKey} to use for looking up the outgredient.
+     * @param fallback The fallback to use if the tag-based lookup did not yield a conclusive outgredient.
+     * @param count    The count to use. Corresponds to {@link ItemStack#getCount()}.
+     */
+    public DefaultedItemTagOutgredient(TagKey<Item> tagKey, Optional<Holder<Item>> fallback, int count) {
+        this(tagKey, fallback, count, DataComponentPatch.EMPTY);
+    }
+
+    /**
+     * @param tagKey   The {@link TagKey} to use for looking up the outgredient.
+     * @param fallback The fallback to use if the tag-based lookup did not yield a conclusive outgredient.
+     * @param count    The count to use. Corresponds to {@link ItemStack#getCount()}.
+     */
+    public DefaultedItemTagOutgredient(TagKey<Item> tagKey, Holder<Item> fallback, int count) {
+        this(tagKey, Optional.of(fallback), count, DataComponentPatch.EMPTY);
+    }
+
+    /**
+     * @param tagKey   The {@link TagKey} to use for looking up the outgredient.
+     * @param fallback The fallback to use if the tag-based lookup did not yield a conclusive outgredient.
+     */
+    public DefaultedItemTagOutgredient(TagKey<Item> tagKey, Optional<Holder<Item>> fallback) {
+        this(tagKey, fallback, 1, DataComponentPatch.EMPTY);
+    }
+
+    /**
+     * @param tagKey   The {@link TagKey} to use for looking up the outgredient.
+     * @param fallback The fallback to use if the tag-based lookup did not yield a conclusive outgredient.
+     */
+    public DefaultedItemTagOutgredient(TagKey<Item> tagKey, Holder<Item> fallback) {
+        this(tagKey, Optional.of(fallback), 1, DataComponentPatch.EMPTY);
+    }
+
+    @Override
+    public ItemStack resolve() {
+        return NeoForgeEventHandler.getTagDefaultsManager()
+                .resolve(Registries.ITEM, tagKey)
+                .map(item -> new ItemStack(item.builtInRegistryHolder(), count, components))
+                .orElseGet(() -> fallback.map(item -> new ItemStack(item, count, components)).orElse(ItemStack.EMPTY));
```

Comment:
This should also likely be cached in the object after first resolve, similar to `Ingredient#getItems()`. Means its can't be a record as you need a lazily loaded cached result field.

When syncing to client, I also sync the result instead of the resolution method to ensure consistency between sides if unsynced things are involved, e.g. config options. So on the client, all *recipe results* would be just item stacks.